adr/ADR-40.md

```diff
@@ -244,7 +245,173 @@ The default interval for PING is 2 minutes.
+representation) - the server will send multiple error protocol messages (one per
+offense) so if multiple subscriptions, you will be able to notify all of them.
+
+For subscriptions, errors are _terminal_ for the subscription, as the server
+cancels the clients interest. so the client will never get any messages on it.
+It is very convenient for client user code to receive an error using some
+mechanism associated with the subscription in question as this will simplify the
+handling of the client code.
+
+It is also useful to have some sort of Promise/Future/etc that will get resolved
+when a subscription closes (will not yield any more messages) - The
+Promise/Future can resolve to an error or void (not thrown) which the client can
+inspect for the reason if any why the subscription closed. Throwing an error is
+discouraged, as this would create a possibility of crashing the client. Clients
+can then use this information to perform their own error handling which may
+require taking the service offline if the subscription is vital for its
+operation.
+
+Note that regardless of a localized error handling mechanism, you should also
+notify the async error handler as you don't know exactly where the client code
+is looking for errors.
+
+#### Authorization Violation
+
+`Authorization Violation` is sent whenever the credentials for a client are not
+accepted. This is followed by a server initiated disconnect. Clients will
+normally reconnect (depending on their connection options). If the client
+closes, this should be reported as the last error.
+
+#### User Authentication Expired
+
+`User Authentication Expired` protocol error happens whenever credentials for
+the client expire while the client is connected to the server. It is followed by
+a server disconnect. This error should be notified in the async handler. On
+reconnect the client is going to be rejected with `Authorization Violation` and
+follow its reconnect logic.
+
+#### Account Expiration
+
+`Account Authentication Expired` is sent whenever the account JWT expires and a
+client for the account is connected. This will result in a disconnect initiated
+by the server. On reconnect the client will be rejected with
+`Authorization Violation` until the account configuration is refreshed on the
+server. The client will follow its reconnect logic.
+
+#### Secure Connection - TLS Required
+
+`Secure Connection - TLS Required` is sent if the client is trying to connect on
+a server that requires TLS.
+
+_????????_ The client should have done extensive ServerInfo investigation and
+determined that this would have been a failure
+
+#### Maximum Number of Connections
+
+`maximum connections exceeded` server limit on number of connections reached.
+Server will send to the client the `-ERR maximum connections exceeded`, client
+possibly go in reconnect loop.
+
+_????????_ The server can also send
+`Connection throttling is active. Please try again later.` when too many TLS
```

Comment:
Add in that this can happen when `tls { connection_rate_limit }` is being used maybe?
https://github.com/nats-io/nats-server/blob/main/server/opts.go#L4557